### ISSUE TITLE
Invalid authentification

### ISSUE THREAD

**Issue by Housemann** [marcelwestrahome/home-assistant-niu-component#57]
I can Not Login to my Niu Account with the integration 
User Passwort and my Scooter id wich Beginns with 199 dont work 
Works the integration?

**Comment by caiodelgadonew** [comment]
Same issue here, but I dont know how to get my scooter id...

**Comment by marcelwestrahome** [comment]
integration works, what kind of error do you get ?. id =0 for one scooter

**Comment by caiodelgadonew** [comment]
@marcelwestrahome I get only invalid authentication 

![CleanShot 2024-07-03 at 19 30 33](https://github.com/marcelwestrahome/home-assistant-niu-component/assets/39803009/1e27eaed-fa69-44fa-9411-1f67ad2b68e0)


Tried with both e-mail and password or username and password. 

Login user/pass is the same as the android NIU app

**Comment by marcelwestrahome** [comment]
in which country do you live ? not sure if integration works for all countries

**Comment by caiodelgadonew** [comment]
@marcelwestrahome netherlands

Btw, there's no option to choose the country

**Comment by marcelwestrahome** [comment]
some countries has different api addresses, but netherlands should work. What kind of scooter do you have and are you really sure about the password ?

**Comment by caiodelgadonew** [comment]
@marcelwestrahome yeah, I use a password manager and even tried to logout login again in the app which works well.

I got a NIU NQi Sport Standard Range. 

Is there any way to enable a verbose/debug level on the integration?

**Comment by caiodelgadonew** [comment]
@marcelwestrahome seems I found the issue... I just generated a new password without special characters and now I can login

**Comment by ralfweber** [comment]
Hello @marcelwestrahome,
is there a way to adapt the code so that passwords with special characters also work? That would also be a good step in terms of security. I would be very happy because then I could also use the integration, but weak passwords without a single special character are not a solution. Are there any solutions here?

**Comment by D3NY75** [comment]
Hello, the integration doesn't work for me either. Whether the password has special characters or not!
I left the ID at 0. Right?
I also logged out on my mobile phone, in case that plays a role.
I would be very grateful for help.

Update: Got it! The username is the email address!

**Comment by florianatgithub** [comment]
Any other Problems known? I don't have a special character in Password, but login don't work. Country is Germany.

**Comment by itsJereMe** [comment]
In my case the problem was that my password was longer than 20 characters

**Comment by Catoputh** [comment]
Does it work in China，and what's meaning of the scooter id?How to
get it? thank you.

**Comment by florianatgithub** [comment]
That was the Solution, Thank You!

**Comment by huajiejun** [comment]
I encountered the same problem, I confirm that the account password is correct, and there are no special characters.

**Comment by huajiejun** [comment]
2024-09-26

Whether it is the overseas version or the domestic modified version, there is a problem of login failure after a few days. It is necessary to restart HA in order to normally pull the latest data. An appropriate solution has not yet been found (scheduled restart of HA should be feasible).

**Comment by Fitosoft** [comment]
> 2024-09-26
> 
> Whether it is the overseas version or the domestic modified version, there is a problem of login failure after a few days. It is necessary to restart HA in order to normally pull the latest data. An appropriate solution has not yet been found (scheduled restart of HA should be feasible).

@huajiejun my PRs should improve stability of user sessions. I hope that helps your issues as well.

**Comment by ionni73** [comment]
> Hello [@marcelwestrahome](https://github.com/marcelwestrahome), is there a way to adapt the code so that passwords with special characters also work? That would also be a good step in terms of security. I would be very happy because then I could also use the integration, but weak passwords without a single special character are not a solution. Are there any solutions here?

Greetings @marcelwestrahome I have a NQiX 150 and looking forward to use your Integration. I live in Germany. Same as the others I had issues in logging in. I don't want to downgrade my password to a weaker one (= no special character). I would wait if the fix is done. Do you have any plans in the future to allow logging in using passwords with special characters? Many thanks for your effort!

**Comment by Fitosoft** [comment]
I can look into the password issue. But may I argue that for random passwords, length matters more than special characters. An alphanumeric password only needs to be roughly 10% longer to match the theoretical strength of one using the full printable character set.

So, for example, 18 random alphanumeric characters are roughly comparable to 16 characters with special symbols. In practice, 20+ random alphanumeric characters generated by a password manager are already extremely strong. NIST also no longer recommends mandatory complexity rules like requiring special characters.

Hopefully that can help you using the integration anyways in the meanwhile. 

Cheers, Tom

**Comment by Fitosoft** [comment]
Hi everyone,

a short update on this issue.

First, I would like to clarify my previous comment about password strength. It was only meant as a temporary workaround, not as a dismissal of the reported problem. Users should not have to change a valid NIU password just to use this integration.

I reviewed the authentication code once more. In the current implementation:

- the password is encoded as UTF-8, including non-ASCII characters such as umlauts
- an MD5 hash is calculated from those UTF-8 bytes, as expected by the NIU login API
- only the hexadecimal hash is sent as password
- the login fields are submitted as standard form-encoded data

Therefore, according to our current understanding, special characters should not be lost or incorrectly escaped by the integration. The code also does not impose a 20-character password limit. However, without any NIU API documentation, we cannot yet confirm whether the NIU backend or app applies additional normalization or length rules.

Several other reports in this discussion are probably separate from the initial password problem:

- The NIU username needs to be the account email address not the self assigned NIU user name.
- The current “scooter ID” field expects a zero-based vehicle index, not the scooter serial number or name. Current error handling seems to distinguish an invalid vehicle selection from invalid authentication.
- Login failures appearing only after several days, which disappeared after restarting Home Assistant, were most likely session-expiration problems rather than invalid credentials.

Current `master` now:
- refreshes tokens, retries after an HTTP 401 response and falls back to a new password login when necessary.
- Authentication, connection, server, rate-limit and invalid-vehicle errors are now reported separately during setup.
- Locale and country handling has also been corrected.

@Housemann @caiodelgadonew @ralfweber @florianatgithub @itsJereMe @huajiejun @ionni73: Would any of you be willing to test the current `master` version manually to see, if any of your issues persist?

If you are comfortable installing a development version, testing `master` would give us the most useful result. Otherwise, please wait for the next release and test again with that version. The currently released version `2.1.8` predates most of the improvements described above, so any test with that version would probably not help us isolate any remaining problems.

If authentication still fails, please include:

- Home Assistant version
- integration version or exact commit
- country and vehicle/scooter model
- whether the failure happens during initial setup or only after some hours/days
- whether the same email address and password work in the official NIU app at that time
- password length and the types of characters used, but of course *never the password itself*
- the exact error shown in Home Assistant
- the relevant Home Assistant log lines - if possible!

Debug logging can be enabled with:

```yaml
logger:
  default: info
  logs:
    custom_components.niu: debug
```

Please remove or mask email addresses, passwords, access tokens and serial numbers before posting logs!

These details would help us determine whether a remaining failure is caused by password handling, a regional NIU API difference, an account-specific backend rule, or session recovery.

Thank you for helping us test this.

**Comment by ionni73** [comment]
Hi @marcelwestrahome,

Many thanks for the explanation and your effort. 

Before starting with the test, I checked that the (old) version of the home assistant integration was still "buggy".
"Unfortunately" it wasn't: I used my credentials (with the password 9 characters long, capital and non-capital letters, letters and numbers and special character) and...it worked.

Happy that it worked, but disappointed that I could not be your tester 🫤

Many thanks, whatever you did!

It works!

Cheers

**Comment by Fitosoft** [comment]
@Housemann @caiodelgadonew @ralfweber @florianatgithub @itsJereMe @huajiejun please let us know here, if you still face any issues with the latest released version (2.1.10 - don't forget to restart HA after updating!). Otherwise I would close this issue after one more week. Of course you can create a new issue anytime - if you do so, please provide as many details as you can. You can find detailed tutorials online on how to retrieve logs. Thanks for your understanding. :)

**Comment by Fitosoft** [comment]
Authentication should be working now. Please report a separate bug with detailed description, if you encounter any issues. Thank you!
